streams/shared/src/main/scala/zio/stream/experimental/ZSink.scala

```diff
@@ -77,7 +77,35 @@ class ZSink[-R, -InErr, -In, +OutErr, +L, +Z](val channel: ZChannel[R, InErr, Ch
   def collectAllWhileWith[S](z: S)(p: Z => Boolean)(f: (S, Z) => S)(implicit
     ev: L <:< In
   ): ZSink[R, InErr, In, OutErr, L, S] =
-    ???
+    new ZSink(
+      ZChannel
+        .fromZIO(Ref.make(Chunk[In]()).zip(Ref.make(false)))
+        .flatMap { case (leftoversRef, upstreamDoneRef) =>
+          lazy val upstreamMarker: ZChannel[Any, InErr, Chunk[In], Any, InErr, Chunk[In], Any] =
+            ZChannel.readWith(
+              (in: Chunk[In]) => ZChannel.write(in) *> upstreamMarker,
+              ZChannel.fail(_: InErr),
+              (x: Any) => ZChannel.fromZIO(upstreamDoneRef.set(true)).as(x)
+            )
+
+          def loop(currentResult: S): ZChannel[R, InErr, Chunk[In], Any, OutErr, Chunk[L], S] =
+            channel.doneCollect
+              .foldChannel(
+                ZChannel.fail(_),
+                { case (leftovers, ze) =>
+                  if (p(ze)) {
+                    for {
+                      _            <- ZChannel.fromZIO(leftoversRef.set(leftovers.flatten.asInstanceOf[Chunk[In]]))
+                      upstreamDone <- ZChannel.fromZIO(upstreamDoneRef.get)
+                      result       <- if (upstreamDone) ZChannel.end(f(currentResult, ze)) else loop(f(currentResult, ze))
```

Comment:
Write the leftovers before terminating:

```suggestion
                      result       <- if (upstreamDone) ZChannel.write(leftovers.flatten) *> ZChannel.end(f(currentResult, ze)) 
                                           else loop(f(currentResult, ze))
```

Reply:
fixed, thanks!